Add test for nodeConnector output cache

SamuPS opened the pull request.

Commit:
Add test for nodeConnector output cache

Commit:
Add same test for rx2 NodeConnector

Commit:
Merge branch 'master' of https://github.com/badoo/RIBs into add-test-to-unloackable

Commit:
Update NodeConnector with thread safe solution.
Update test to match new NodeConnector.

Commit:
Suprress FunctionNaming detekt error for BaseTestClases

Commit:
Remove duplicated test helpers

Commit:
Merge remote-tracking branch 'zsoltk/add-unlockable' into add-test-to-unloackable

Commit:
Reduce Rx2NodeConnectorTest executions to 1000 as with the duplicated logic that is enougth.

Commit:
Update NodeConnector Error Message

Commit:
Fix typo in test comment

Commit:
Improve test comment

Commit:
Improve test comment in Rx2Test

CherryPerry commented on libraries/rib-base-test/src/main/java/com/badoo/ribs/test/BaseConnectableNodeTest.kt:
```diff
@@ -0,0 +1,13 @@
+package com.badoo.ribs.test
+
+@Suppress("FunctionNaming")
+abstract class BaseConnectableNodeTest {
```
You leak unnecessary utility class to library consumers.
I think it is easier to duplicate test methods rather than implement separate utility Gradle module for it.

zsoltk replied:
This was my suggestion, the point being not just to avoid copy-paste, but to actually enforce same expectations (test suite) for all implementations.

SamuPS replied:
Do you have any suggestions about how to use the abstraction approach without exposing it? If the base-test is not supossed to be exposed, i would go for removing the abstract or to create an internal utils module, but not sure how desirable is that.

zsoltk replied:
I don't think it's a problem to expose it. We also can't provide all possible implementations of `Connectable`, so it might even help to have this test suite definition openly available.

CherryPerry commented on libraries/rib-base-test/build.gradle:
```diff
@@ -31,6 +31,7 @@ dependencies {
     implementation deps.android.runner
     implementation deps.test.junit4
     implementation deps.test.mockito
+    implementation deps.test.mockitoKotlin
```
See comment below and remove.

CherryPerry commented on libraries/rib-base-test/src/main/java/com/badoo/ribs/test/BaseConnectableNodeTest.kt:
```diff
@@ -0,0 +1,13 @@
+@Suppress("FunctionNaming")
+abstract class BaseConnectableNodeTest {
+
+    abstract fun WHEN_child_emit_some_output_before_it_is_attached_to_parent_THEN_parent_receive_the_output_after_child_attach_finished()
```
You can use Kotlin names here fun `WHEN child emit...`()

CherryPerry commented on libraries/rib-base-test/src/main/java/com/badoo/ribs/test/helper/connectable/ConnectableTestNode.kt:
```diff
@@ -0,0 +1,44 @@
+import com.badoo.ribs.test.helper.testBuildParams
+import com.nhaarman.mockitokotlin2.mock
+
+class ConnectableTestNode(
```
Same here.
rib-base-test it not an internal utility module, but publicity available one that is intended to be used.

SamuPS replied:
I will move it away, it was added here to avoid some interface naming conflict. But if this is the case, let's rename the interface instead

CherryPerry commented on libraries/rib-base/src/test/java/com/badoo/ribs/clienthelper/connector/NodeConnectorTest.kt:
```diff
@@ -0,0 +1,246 @@
+        val executor = Executors.newFixedThreadPool(threadNumber).apply {
+            repeat(threadNumber) {
+                submit {
+                    barrier.await()
```
Specify timeout for every `await` you use to avoid dead locks (locally and on CI). Like extension function with 1 minute default timeout.

CherryPerry commented on libraries/rib-base/src/test/java/com/badoo/ribs/clienthelper/connector/NodeConnectorTest.kt:
```diff
@@ -0,0 +1,246 @@
+        nodeConnector.output.accept(Output2)
+        nodeConnector.output.accept(Output3)
+
+        firstTestObserver.assertValueAt(0, Output1)
```
`firstTestObserver.assertValues(Output1, Output2, Output3)`
> Assert that the TestObserver/TestSubscriber received only the specified values in the specified order.

CherryPerry commented on libraries/rib-base/src/test/java/com/badoo/ribs/clienthelper/connector/NodeConnectorTest.kt:
```diff
@@ -0,0 +1,246 @@
+    @Test
+    override fun GIVEN_nodeConnector_onAttached_is_not_called_WHEN_output_is_accepted_THEN_accepted_output_do_not_reach_observer() {
+        val nodeConnector = NodeConnector<Input, Output>()
+        nodeConnector.output.observe { firstTestObserver.onNext(it) }
```
`nodeConnector.output.observe(firstTestObserver)`

SamuPS replied:
Not possible for this ones, but i would do it for the Rx2 ones.

CherryPerry commented on libraries/rib-rx2/src/test/java/com/badoo/ribs/rx2/clienthelper/conector/helpers/Rx2ConnectableTestNode.kt:
```diff
@@ -0,0 +1,43 @@
+    buildParams: BuildParams<*>? = null,
+    parent: Node<*>? = null,
+    viewFactory: ViewFactory<TestView>? = null,
+    router: Router<*> = mock(),
```
Try to run without router.

SamuPS replied:
Removed, i also removed other non required elements.

Commit:
Fix execution order in test

Commit:
Remove testConfiguration from rib-base
Move ConnectableTestNode from rib-base
Remove unrequired elements fomr ConnectableTestNode
Use space instead of underscores for BaseConnectableNodeTest
Add barrier await timeOut
Improve test asserts

Commit:
Fix test

CherryPerry commented on libraries/rib-base/src/main/java/com/badoo/ribs/core/plugin/Plugin.kt:
```diff
@@ -21,7 +21,7 @@ interface NodeAware : Plugin {
 
 interface NodeLifecycleAware : Plugin {
-    fun onBuild() {}
+    fun onNodeBuild() {}
```
Why did you rename it?

SamuPS replied:
Otherwise no class extending Node defined inside rib-base can extend connectables, as there is an interfaces clash(Node onBuild is internal).

SamuPS replied:
Basically this has to do with taking away ConnectableTestNode back to rib-base. Maybe we do need a new module for internal test utility.

CherryPerry replied:
Renaming internal method seems like more easy way, it will affect only one file.

I start thinking, that it was a bad idea to let Node implement Plugins (because of name clash for example). Simple check inside `node.onAttach()` like `if (this is Connectable) this.unlock()` maybe was not that bad.

SamuPS replied:
Well, the scenario for this clash shouldn't be a concern for the library users isn't it? I find pretty flexible the idea of Node implementing the plugings, but also my understanding it not as deep as your so, open to go this way.

zsoltk replied:
Let's keep plugin method name the same and change the internal method.

Commit:
Rename Node instead of plugin

Commit:
Rename internal Node onBuild method instead of plugin interface. Fix some plugin interfaces naming.

Commit:
Change onAttachedChildView to onChildViewAttached

Commit:
Attach plugins in MenuRibTest

Commit:
Remove underscore name from BaseNodeConnectorTest

Commit:
Fix SubtreeViewChangeAware plugin naming

Commit:
Remove undesired line from test
